### DISCUSSION TITLE
ESP32 + pms5003 + MQTT

### DISCUSSION THREAD

**Discussion by TomasTnr** [micropython/micropython#14313] [ESP32]
Hi Guys,
I am new in Micropython and I need learn a lot yet but i now am busy with project at home to use ESP32 and PMS5003 to send via mqtt data to Home Assistant. 
I could use Arduino code but i want to learn micropython. 

Til now i used librabry pms5003.py and main.py as follows:

```
import pms5003
import machine
import uasyncio as asyncio

uart = machine.UART(1, tx=25, rx=26, baudrate=9600)
pm = pms5003.PMS5003(uart)
pm.registerCallback(pm.print)

loop=asyncio.get_event_loop()
loop.run_forever()
```
Everyth works as it should but then i wrote a code to send this data via MQTT 

```
import machine
import pms5003
import uasyncio as asyncio
from umqtt.simple import MQTTClient

# Initialize MQTT parameters
mqtt_broker = "xxx.xxx.xxx.xxx"  # Your MQTT broker IP address
mqtt_port = 1883  # Your MQTT broker port
mqtt_topic = "pms5003"  # Topic to publish the sensor readings
mqtt_client_id = "pms5003_client"  # Client ID for MQTT connection
mqtt_username = "XXXXXXXXXXXX"  # MQTT username
mqtt_password = "XXXXXXXXXXXX"  # MQTT password

# Initialize UART for PMS5003 sensor
uart = machine.UART(1, tx=25, rx=26, baudrate=9600)
pm = pms5003.PMS5003(uart)

# Initialize MQTT client
mqtt_client = MQTTClient(mqtt_client_id, mqtt_broker, mqtt_port, mqtt_username, mqtt_password)

async def read_sensor_and_publish():
    while True:
        data = await pm.read()
        if data is not None:
            # Publish sensor readings to MQTT topic
            mqtt_payload = "PM1.0: {}, PM2.5: {}, PM10: {}".format(data[0], data[1], data[2])
            mqtt_client.publish(mqtt_topic, mqtt_payload)
        await uasyncio.sleep(60)  # Publish readings every 60 seconds

async def main():
    # Connect to MQTT broker
    mqtt_client.connect()
    # Start reading sensor and publishing readings to MQTT
    uasyncio.create_task(read_sensor_and_publish())

pm.registerCallback(pm.print)

# Run the main coroutine
# loop=asyncio.get_event_loop()
# loop.run_forever()
uasyncio.run(main())
```

I know someth is wrong but i don't know what and i think that my little knowledge is the issue ;)
Can someone help me with this?

PS. I have code to connect to WIFI in boot.py

**Comment by peterhinch** [comment]
Your use of `asyncio` is incorrect. Firstly with recent firmware you can forget about `uasyncio`. Just import `asyncio` and use it.

The problem is that `main` will run to completion (`.create_task` returns immediately). Once `main` has terminated, `asyncio.run` will terminate and the application will quit, returning to the REPL before it has a chance to do anything useful. I would do something like this:
```py
async def main():
    mqtt_client.connect()
    while True:
        data = await pm.read()
        if data is not None:
            # Publish sensor readings to MQTT topic
            mqtt_payload = "PM1.0: {}, PM2.5: {}, PM10: {}".format(data[0], data[1], data[2])
            mqtt_client.publish(mqtt_topic, mqtt_payload)
        await uasyncio.sleep(60)  # Publish readings every 60 seconds

pm.registerCallback(pm.print)
asyncio.run(main())
```
This assumes that `pm.read()` is an asynchronous function - I'm unfamiliar with the library.

Resources which may be of interest: [asyncio tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md). Once you've got MQTT working you might want to look at [asynchronous MQTT](https://github.com/peterhinch/micropython-mqtt).
